implementations/rust/ffi/src/kex.rs

```diff
@@ -0,0 +1,210 @@
+use crate::error::{Error, FfiOckamError};
+use crate::kex_types::*;
+use crate::mutex_storage::FfiObjectMutexStorage;
+use crate::vault::{DEFAULT_VAULTS, FILESYSTEM_VAULTS, SECRETS};
+use crate::vault_types::{FfiVaultFatPointer, FfiVaultType};
+use ockam_kex::{CipherSuite, KeyExchanger};
+use ockam_kex_xx::{SymmetricState, XXInitiator, XXResponder, XXVault};
+use std::ops::DerefMut;
+use std::sync::{Arc, Mutex};
+
+lazy_static! {
+    pub(crate) static ref XX_INITIATOR: FfiObjectMutexStorage<XXInitiator> =
+        FfiObjectMutexStorage::default();
+    pub(crate) static ref XX_RESPONDER: FfiObjectMutexStorage<XXResponder> =
+        FfiObjectMutexStorage::default();
+}
+
+fn call<F, R>(context: FfiKexFatPointer, callback: F) -> Result<R, FfiOckamError>
+where
+    F: FnOnce(&mut dyn KeyExchanger) -> Result<R, FfiOckamError>,
+{
+    match context.kex_type {
+        FfiKexType::XxInitiator => {
+            let item = XX_INITIATOR.get_object(context.handle)?;
+            let mut item = item.lock().unwrap();
+
+            callback(item.deref_mut())
+        }
+        FfiKexType::XxResponder => {
+            let item = XX_RESPONDER.get_object(context.handle)?;
+            let mut item = item.lock().unwrap();
+
+            callback(item.deref_mut())
+        }
+    }
+}
+
+fn cast_vault(vault: FfiVaultFatPointer) -> Result<Arc<Mutex<dyn XXVault>>, FfiOckamError> {
+    let vault: Arc<Mutex<dyn XXVault>> = match vault.vault_type {
+        FfiVaultType::Software => DEFAULT_VAULTS.get_object(vault.handle)?,
+        FfiVaultType::Filesystem => FILESYSTEM_VAULTS.get_object(vault.handle)?,
+    };
+
+    Ok(vault)
+}
+
+/// Create a new kex initiator and return it
+#[no_mangle]
+pub extern "C" fn ockam_kex_xx_initiator(
+    context: &mut FfiKexFatPointer,
+    vault: FfiVaultFatPointer,
+    identity_key: u64,
+) -> FfiOckamError {
+    let res = || {
```

Comment:
I don't understand this. Why are you creating an anonymous function then immediately calling it?
What's the purpose of the anonymous function?

Reply:
The purpose is to allow using ? operator inside